New Blog Post: SimpleChains.jl Specializing Scientific Machine Learning with Julia (#1661)
JuliaLang/www.julialang.org: blog/2022/04/simple-chains.md

```
Took: 17.33
Accuracy: 0.9492
Took: 17.22
Accuracy: 0.9751
```
PyTorch using an AMD EPYC 7513 32-Core Processor:
```
Took: 14.86
Accuracy: 0.9626
Took: 15.09
Accuracy: 0.9783
```
PyTorch using an Intel i9 10980XE 18-Core Processor:
```
Took: 11.24
Accuracy: 0.9759
Took: 10.78
Accuracy: 0.9841
```

#### Flux.jl

The standard machine learning library in Julia, [Flux.jl](https://github.com/FluxML/Flux.jl) was benchmarked using [this script]() with an A100 GPU. How fast was that?

```julia
julia> @time train!(model, train_loader)
 74.678251 seconds (195.36 M allocations: 12.035 GiB, 4.28% gc time, 77.57% compilation time)

julia> eval_loss_accuracy(train_loader, model, device),
       eval_loss_accuracy(test_loader, model, device)
((loss = 0.1579f0, acc = 95.3583), (loss = 0.1495f0, acc = 95.54))

julia> @time train!(model, train_loader)
  1.676934 seconds (1.04 M allocations: 1.840 GiB, 5.64% gc time, 0.63% compilation time)

julia> eval_loss_accuracy(train_loader, model, device),
       eval_loss_accuracy(test_loader, model, device)
((loss = 0.0819f0, acc = 97.4967), (loss = 0.076f0, acc = 97.6))

> ```

Reviewer: Would help for understanding/readability if made clear that the first run includes compile time.  It is kind of mentioned in the summary but readers may get confused if not mentioned earlier.

Reviewer 2: I'll update these examples tonight.
1. Timings will change to reflect the latest changes.
2. I'll add a note about compile time, and also try to improve SimpleChains'.
3. I'll switch the acc, loss order for the Flux benchmarks to reflect the order for SimpleChains and PyTorch.

Reviewer 2: I'll also add a small mlp example.

(0.9744, 0.08624289f0)

julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
  3.083624 seconds

julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
(0.9835666666666667, 0.056287352f0)

julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
(0.9831, 0.053463124f0)
```

SimpleChains on an Intel i9 10980XE 18-Core Processor:

```julia
julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
 34.915917 seconds (83.99 M allocations: 5.431 GiB, 3.91% gc time, 95.22% compilation time)

julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
(0.972, 0.09354472f0)

julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
(0.975, 0.084895514f0)

julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
  1.324148 seconds

julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
(0.9836833333333334, 0.054299146f0)

julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
(0.9829, 0.053510584f0)
```

SimpleChains on an Intel i7 1165G7 4-Core Processor (thin and light laptop CPU):

```julia

> julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
>  42.664702 seconds (103.71 M allocations: 5.236 GiB, 2.85% gc time, 77.60% compilation time)
> 
> julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
> (0.9863166666666666, 0.044742946f0)
> 
> julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
> (0.9841, 0.048268355f0)
> 
> julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
>   6.320631 seconds
> 
> julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
> (0.99485, 0.017716348f0)
> 
> julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
> (0.9888, 0.035394628f0)

Reviewer 2: ```suggestion
julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain, SimpleChains.ADAM(3e-4), 10);
 41.053800 seconds (104.10 M allocations: 5.263 GiB, 2.83% gc time, 77.62% compilation time)

julia> SimpleChains.accuracy_and_loss(lenetloss, xtrain, p),
       SimpleChains.accuracy_and_loss(lenetloss, xtest, ytest, p)
((0.9491333333333334, 0.16993132f0), (0.9508, 0.15890576f0))

julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain, SimpleChains.ADAM(3e-4), 10);
  5.320512 seconds

julia> SimpleChains.accuracy_and_loss(lenetloss, xtrain, p),
       SimpleChains.accuracy_and_loss(lenetloss, xtest, ytest, p)
((0.9700833333333333, 0.10100537f0), (0.9689, 0.09761506f0))
```

Reviewer 2: My poor laptop has a hard time taking the heat, runs out of power budget, or something that causes it to begin throttling.
So I didn't copy/paste the script as one giant block, but waited a few seconds before running the second `@time SimpleChains.train_batched!` to let it cool off. Might be worth mentioning this.

```

Note that smaller batch sizes improve accuracy per epoch, and batch sizes were set to be proportional to the number of threads.

## Benchmark Summary

Latency before the first epoch begins training is problematic, but SimpleChains.jl is fast once compiled.

> Post-compilation, the 10980XE was around 70% faster than Flux with the A100 GPU.

Reviewer 2: ```suggestion
Post-compilation, the 10980XE was competitive with Flux using an A100 GPU, and about 35% faster than the V100.
```
I realized the loaders weren't putting the data onto the GPU up front. Doing this improved the Flux GPU times (in the suggestions above), so my advantage is smaller.

But that makes Flux look quite good compared to PyTorch on the GPU, which was oddly slow.
This also suggests that PyTorch's problem isn't that the data wasn't already on the GPU, as Flux was many times faster when experiencing that same problem.

The 1165G7, a laptop CPU featuring AVX512, was competive, handily trouncing any of the competing machine learning libraries when 

> they were run on CPU, and even beat PyTorch on a V100. Again, we stress that this test case followed the more typical machine learning

Reviewer 2: ```suggestion
they were run on far beefier CPUs, and even beat PyTorch on both the V100 and A100. Again, we stress that this test case followed the more typical machine learning
```

@def rss_pubdate = Date(2022, 4, 11)
@def rss = """10x-ing PyTorch: Specializing Scientific Machine Learning with Julia ..."""
@def published = "11 April 2022"

> @def title = "10x-ing PyTorch: Specializing Scientific Machine Learning with Julia"

Reviewer 3: I'm not sure about the 10x claim when the benchmarks in the post does not quite support that.

Author: 11.24 seconds to 1.324148 seconds?

Reviewer 3: That could be close enough for some "artistic licence" but it also looks a bit cherry-picked as the AMD benchmark is only 5x faster. 

As it stands now, I think the Julia community would not bat an eye but I'd expect a bit of eye-rolling from those feeling a little defensive of PyTorch. 10x-ing is a better punchphrase than something softer or with qualifiers though. @chriselrod would we support our case by adding a more SciML-representative benchmark of a small MLP?

Author: We can go with 5x, though we know the benchmark is ultra-conservative given the showcase of how matvecs will have a much larger difference, and the benchmark is matmuls.

Reviewer 2: > 11.24 seconds to 1.324148 seconds?

I'll try a couple things to see if I can squeeze out another 20%.

Reviewer 2: Okay, first thing I tried was about a 2% improvement.

My other idea will require a more radical change, which is transposing the internal data so we can SIMD across batch sizes.

Reviewer 2: Do we have any sort of UDE type example using PyTorch, where we can show we >= 10x them?

It'd be great to have two examples:
1. The intended use case, where hopefully we vastly outperform them.
2. Then we show scaling by moving more towards their domain with the PyTorch example, where we can demonstrate that we were so far ahead in the first, we still have plenty of advantage left over with >= 4x.

Reviewer 2: Unfortunately, I think it'd be difficult to extract much more performance out of this example without larger changes.

SIMD-ing across batch may still be worth it.

Author: The problem is that to full specialize the UDE training on SimpleChains, SimpleChains needs to be open sourced so we can integrate it into DiffEqSensitivity.jl and such, which is a chunk of what will be done next 😅.  So I think we just have to choose a stopping point somewhere. If it's easy, try doing the spiral neural ODE (https://gist.github.com/ChrisRackauckas/4a4d526c15cc4170ce37da837bfc32c4) which is from the neural ODE paper.

Reviewer 3: Going full-out SciML and using UDEs in the benchmark might be problematic anyways since much of the time difference could be in the DE solvers. I was more thinking of something like training an MLP to map from a vector of 10 inputs to a single output. That seems like it could be pretty "clean" in the sense that it really only benchmarks the ML tool and representative in that it's the kind of networks we often repeatedly solve in UDEs.

> @def authors = """Chris Elrod, Chris Rackauckas"""

Reviewer 4: ```suggestion
@def authors = """Chris Elrod, Niklas Korsbo, Chris Rackauckas"""
```

```julia
using LinearAlgebra, BenchmarkTools, CUDA, LoopVectorization

function mybroadcastturbo!(C, A, B)
    @tturbo for k ∈ axes(A, 2)
        C[k] += A[k] * B[k]
    end
end

function alloc_timer(n)
    A = rand(Float32,n,n)
    B = rand(Float32,n,n)
    C = rand(Float32,n,n)
    t1 = @belapsed $A .* $B
    t2 = @belapsed ($C .= $A .* $B)
    t3 = @belapsed (mybroadcastturbo!($C, $A, $B))
    A,B,C = (cu(A), cu(B), cu(C))
    t4 = @belapsed CUDA.@sync($A .* $B)
    t5 = @belapsed CUDA.@sync($C .= $A .* $B)
    t1,t2,t3,t4,t5
end
ns = 2 .^ (2:11)
res = [alloc_timer(n) for n in ns]
alloc      = [x[1] for x in res]
noalloc    = [x[2] for x in res]
noalloclv  = [x[3] for x in res]
allocgpu   = [x[4] for x in res]
noallocgpu = [x[5] for x in res]

using Plots
plot(ns,alloc,label="=",xscale=:log10,yscale=:log10,legend=:bottomright,
     title="Which Micro-optimizations matter for BLAS1?",
     ylabel = "Time (s)", xlabel = "N",
     yticks = 10.0 .^ (-8:0.5:2),)
plot!(ns,noalloc,label=".=")

> plot!(ns, noalloc, label="mybroadcastturbo!")

Reviewer 3: The label here does not match that of the plot.

1. Allow us to achieve good performance, ideally approaching the CPU's potential peak FLOPs.
2. Focus on small size meant we could largely forgo large kernel optimizations (such as cache tiling) in the early stages of development.
3. Have an API where vectors of parameters (and their gradients) are first class, rather than having parameters live with the layers, to make it easier to work with various optimizers or solvers that expect contiguous vectors (such as BFGS).
4. Be written in "pure Julia" for ease of development and optimization; while making heavy use of [LoopVectorization.jl](https://github.com/JuliaSIMD/LoopVectorization.jl/), SimpleChains.jl does not rely on any BLAS or NN libraries. It is a long term aim to extend this loop-compiler approach to optimization to also producing pullbacks automatically, without requiring them to be handwritten. However, the compiler-focused approach is already levered for ease of implementation: while we still have to hand-write gradients, we do not need to hand-optimize them.

> ## SimpleChains.jl in Action: 10x-ing PyTorch in Small Examples

Reviewer 3: Well, the benchmarks here are not quite 10x. Some of them a fair bit less than 10x.

  SimpleChains.Conv(SimpleChains.relu, (5, 5), 6),
  SimpleChains.MaxPool(2, 2),
  SimpleChains.Conv(SimpleChains.relu, (5, 5), 16),
  SimpleChains.MaxPool(2, 2),
  Flatten(3),
  TurboDense(SimpleChains.relu, 120),
  TurboDense(SimpleChains.relu, 84),
  TurboDense(identity, 10),
)

# 3d and 0-indexed
xtrain3, ytrain0 = MLDatasets.MNIST.traindata(Float32);
xtest3, ytest0 = MLDatasets.MNIST.testdata(Float32);

xtrain4 = reshape(xtrain3, 28, 28, 1, :);
xtest4 = reshape(xtest3, 28, 28, 1, :);

ytrain1 = UInt32.(ytrain0 .+ 1);
ytest1 = UInt32.(ytest0 .+ 1);

lenetloss = SimpleChains.add_loss(lenet, LogitCrossEntropyLoss(ytrain1));

# initialize parameters
@time p = SimpleChains.init_params(lenet);

# initialize a gradient buffer matrix; number of columns places an upper bound
# on the number of threads used.
G = similar(p, length(p), min(Threads.nthreads(), (Sys.CPU_THREADS ÷ ((Sys.ARCH === :x86_64) + 1))));

# train
@time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);

# assess training and test loss
SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)

> 
> 
> # reinitialize to randomize parameters
> # SimpleChains.init_params!(lenet, p);

Reviewer 3: ```suggestion
```

> # train without additional memor allocations

Reviewer 3: ```suggestion
# train without additional memory allocations
```

Do these assumptions actually hold in our case? And if they don't, can we focus on these aspects to draw more performance out for our use cases?

## Digging In: Small Neural Network Performance Overheads

It's easy to show that these assumptions breakdown when we start focusing on this smaller neural network use case. For starters, lets look at assumptions (1)(2). It's not hard to show where these two will unravel:

```julia
using LinearAlgebra, BenchmarkTools, CUDA, LoopVectorization

function mygemmturbo!(C, A, B)
    @tturbo for m ∈ axes(A, 1), n ∈ axes(B, 2)
        Cmn = zero(eltype(C))
        for k ∈ axes(A, 2)
            Cmn += A[m, k] * B[k, n]
        end
        C[m, n] = Cmn
    end
end

function alloc_timer(n)
    A = rand(Float32,n,n)
    B = rand(Float32,n,n)
    C = rand(Float32,n,n)
    t1 = @belapsed $A * $B
    t2 = @belapsed (mul!($C,$A,$B))
    t3 = @belapsed (mygemmturbo!($C,$A,$B))
    A,B,C = (cu(A), cu(B), cu(C))
    t4 = @belapsed CUDA.@sync($A * $B)
    t5 = @belapsed CUDA.@sync(mul!($C,$A,$B))
    t1,t2,t3,t4,t5
end
ns = 2 .^ (2:11)
res = [alloc_timer(n) for n in ns]

> alloc      = [x[1] for x in res]
> noalloc    = [x[2] for x in res]
> noalloclv  = [x[3] for x in res]
> allocgpu   = [x[4] for x in res]
> noallocgpu = [x[5] for x in res]

Reviewer: ```suggestion
alloc      = [t[1] for t in res]
noalloc    = [t[2] for t in res]
noalloclv  = [t[3] for t in res]
allocgpu   = [t[4] for t in res]
noallocgpu = [t[5] for t in res]
```

Matrix-matrix operations only occur when batching is able to be used (where each column of the B matrix in A*B is a separate batch). In many cases in scientific machine learning, such as [the calculation of vector-Jacobian products in ODE adjoints](https://youtu.be/6hhF6Llv4sI?t=342), this operation is a matrix-vector multiplication. These operations are smaller and only O(n^2), and as you would guess these effects are amplified in this scenario:

```julia
using LinearAlgebra, BenchmarkTools, CUDA, LoopVectorization

function mygemmturbo!(C, A, B)
    @tturbo for m ∈ axes(A, 1), n ∈ axes(B, 2)
        Cmn = zero(eltype(C))
        for k ∈ axes(A, 2)
            Cmn += A[m, k] * B[k, n]
        end
        C[m, n] = Cmn
    end
end

function alloc_timer(n)
    A = rand(Float32,n,n)
    B = rand(Float32,n)
    C = rand(Float32,n)
    t1 = @belapsed $A * $B
    t2 = @belapsed (mul!($C,$A,$B))
    t3 = @belapsed (mygemmturbo!($C,$A,$B))
    A,B,C = (cu(A), cu(B), cu(C))
    t4 = @belapsed CUDA.@sync($A * $B)
    t5 = @belapsed CUDA.@sync(mul!($C,$A,$B))
    t1,t2,t3,t4,t5
end
ns = 2 .^ (2:11)
res = [alloc_timer(n) for n in ns]

> alloc      = [x[1] for x in res]
> noalloc    = [x[2] for x in res]
> noalloclv  = [x[3] for x in res]
> allocgpu   = [x[4] for x in res]
> noallocgpu = [x[5] for x in res]

Reviewer: ```suggestion
alloc      = [t[1] for t in res]
noalloc    = [t[2] for t in res]
noalloclv  = [t[3] for t in res]
allocgpu   = [t[4] for t in res]
noallocgpu = [t[5] for t in res]
```

    ylabel="Time (s)", xlabel="N",)
plot!(ns,noalloc,label="mul! (OpenBLAS)")
plot!(ns,noalloclv,label="mygemvturbo!")
plot!(ns,allocgpu,label="* gpu")
plot!(ns,noallocgpu,label="mul! gpu")
savefig("microopts_blas2.png")
```

![](https://user-images.githubusercontent.com/1814174/162625320-310d633a-34bf-407e-8cc9-ec55ca895d83.png)

And remember, the basic operations of a neural network are `sigma.(W*x .+ b)`, and thus there's also an O(n) element-wise operation. As you would guess, this operation becomes more significant as n gets smaller while requiring even more consideration for memory operations. 

```julia
using LinearAlgebra, BenchmarkTools, CUDA, LoopVectorization

function mybroadcastturbo!(C, A, B)
    @tturbo for k ∈ axes(A, 2)
        C[k] += A[k] * B[k]
    end
end

function alloc_timer(n)
    A = rand(Float32,n,n)
    B = rand(Float32,n,n)
    C = rand(Float32,n,n)
    t1 = @belapsed $A .* $B
    t2 = @belapsed ($C .= $A .* $B)
    t3 = @belapsed (mybroadcastturbo!($C, $A, $B))
    A,B,C = (cu(A), cu(B), cu(C))
    t4 = @belapsed CUDA.@sync($A .* $B)
    t5 = @belapsed CUDA.@sync($C .= $A .* $B)
    t1,t2,t3,t4,t5
end
ns = 2 .^ (2:11)
res = [alloc_timer(n) for n in ns]

> alloc      = [x[1] for x in res]
> noalloc    = [x[2] for x in res]
> noalloclv  = [x[3] for x in res]
> allocgpu   = [x[4] for x in res]
> noallocgpu = [x[5] for x in res]

Reviewer: ```suggestion
alloc      = [t[1] for t in res]
noalloc    = [t[2] for t in res]
noalloclv  = [t[3] for t in res]
allocgpu   = [t[4] for t in res]
noallocgpu = [t[5] for t in res]
```

see from a user that is not digging into internals and hyper-optimizing. Nothing out of the ordinary was done with the Julia scripts: these were
all "by the book" implementations. Though of course, the SimpleChains.jl's simplest code is specifically optimized for this "by the book" use case.

## Conclusion

There are many things that can make a library achieve high-performance, and nothing is as essential as knowing how it will be used. While the big machine learning frameworks have done extremely well focusing on the top-notch performance for 99.9% of their users, one can still completely outclass them when focusing on some of the 0.1% of applications which fall outside of what they have been targeting. This is the advantage of composability and flexibility: a language that allows you to easily build a machine learning framework is also a language which allows you to build alternative frameworks which are optimized for alternative people. SimpleChains.jl will not be useful to everybody, but it will be extremely useful to those who need it.

## Opportunities

> If you are interested in this topic and want to work on Julia and machine learning, note that the [DeepPumas team is hiring](https://pumas.ai/company/machine-learning-scientist/). Additionally, [SciML internships are also available at Julia Computing](https://jobs.juliacomputing.com/jobs/ndvlJz9fHYcr/machine-learning-intern-remote).

Reviewer 3: ```suggestion
If you are interested in this topic and want to work on Julia and machine learning, note that the DeepPumas team is hiring ([here](https://pumas.ai/company/machine-learning-scientist/) and [here](https://pumas.ai/company/scientist-deeppumas/)). Additionally, [SciML internships are also available at Julia Computing](https://jobs.juliacomputing.com/jobs/ndvlJz9fHYcr/machine-learning-intern-remote).
```
